File: barberia/lib/core/utils/time_slot_generator.dart
import '../../core/constants/service_duration_constants.dart';
import '../../core/utils/appointment_slot_utils.dart';
import '../../data/models/schedule_config.dart';

enum TimeSlotStatus { available, pastUnavailable, blocked, booked }

class TimeSlotEntry {
  const TimeSlotEntry({
    required this.time,
    required this.status,
  });

  final String time;
  final TimeSlotStatus status;

  bool get isSelectable => status == TimeSlotStatus.available;
}

class TimeSlotGenerator {
  static List<String> generateAllSlots(ScheduleConfig config) {
    return generateSlots(
      start: config.startTime,
      end: config.endTime,
      intervalMinutes: config.intervalMinutes,
    );
  }

  static List<String> generateSlots({
    required String start,
    required String end,
    required int intervalMinutes,
  }) {
    final slots = <String>[];
    var current = _parseTime(start);
    final endMinutes = _parseTime(end);

    while (current <= endMinutes) {
      slots.add(_formatTime(current));
      current += intervalMinutes;
    }

    return slots;
  }

  static List<String> getAvailableSlots(
    List<String> bookedTimes, {
    required ScheduleConfig config,
    required DateTime date,
    List<String> blockedTimes = const [],
    DateTime? reference,
  }) {
    return buildSelectableGrid(
      config: config,
      date: date,
      bookedTimes: bookedTimes,
      blockedTimes: blockedTimes,
      reference: reference,
    )
        .where((entry) => entry.isSelectable)
        .map((entry) => entry.time)
        .toList();
  }

  static List<TimeSlotEntry> buildSelectableGrid({
    required ScheduleConfig config,
    required DateTime date,
    required List<String> bookedTimes,
    List<String> blockedTimes = const [],
    DateTime? reference,
  }) {
    final now = reference ?? DateTime.now();
    final day = DateTime(date.year, date.month, date.day);
    final today = DateTime(now.year, now.month, now.day);
    final isToday =
        day.year == today.year && day.month == today.month && day.day == today.day;

    final entries = <TimeSlotEntry>[];
    final blockedSet = blockedTimes.toSet();

    for (final slot in generateAllSlots(config)) {
      if (bookedTimes.contains(slot)) {
        entries.add(TimeSlotEntry(
          time: slot,
          status: TimeSlotStatus.booked,
        ));
        continue;
      }

      if (blockedSet.contains(slot)) {
        entries.add(TimeSlotEntry(
          time: slot,
          status: TimeSlotStatus.blocked,
        ));
        continue;
      }

      if (isToday && _slotDateTime(day, slot).isBefore(now)) {
        entries.add(TimeSlotEntry(
          time: slot,
          status: TimeSlotStatus.pastUnavailable,
        ));
      } else {
        entries.add(TimeSlotEntry(
          time: slot,
          status: TimeSlotStatus.available,
        ));
      }
    }

    return entries;
  }

  static List<TimeSlotEntry> buildBookingGrid({
    required ScheduleConfig config,
    required DateTime date,
    required List<String> occupiedSlots,
    required int durationMinutes,
    List<String> blockedTimes = const [],
    DateTime? reference,
  }) {
    final bookingConfig = config.copyWith(
      intervalMinutes: ServiceDurationConstants.blockMinutes,
    );
    final occupiedSet = occupiedSlots.toSet();
    final blockedSet = blockedTimes.toSet();
    final now = reference ?? DateTime.now();
    final day = DateTime(date.year, date.month, date.day);
    final today = DateTime(now.year, now.month, now.day);
    final isToday =
        day.year == today.year && day.month == today.month && day.day == today.day;

    final entries = <TimeSlotEntry>[];

    for (final slot in generateAllSlots(bookingConfig)) {
      final slotOccupied = occupiedSet.contains(slot);
      final slotBlocked = blockedSet.contains(slot);
      final isPast =
          isToday && _slotDateTime(day, slot).isBefore(now);

      final fits = AppointmentSlotUtils.canFitAtStart(
        startTime: slot,
        durationMinutes: durationMinutes,
        scheduleEnd: config.endTime,
        occupiedSlots: occupiedSet,
        blockedSlots: blockedSet,
      );

      if (!fits) {
        if (slotOccupied) {
          entries.add(TimeSlotEntry(time: slot, status: TimeSlotStatus.booked));
        } else if (slotBlocked) {
          entries.add(TimeSlotEntry(time: slot, status: TimeSlotStatus.blocked));
        } else if (isPast) {
          entries.add(
            TimeSlotEntry(time: slot, status: TimeSlotStatus.pastUnavailable),
          );
        } else {
          entries.add(
            TimeSlotEntry(time: slot, status: TimeSlotStatus.pastUnavailable),
          );
        }
        continue;
      }

      if (isPast) {
        entries.add(TimeSlotEntry(time: slot, status: TimeSlotStatus.pastUnavailable));
      } else {
        entries.add(TimeSlotEntry(time: slot, status: TimeSlotStatus.available));
      }
    }

    return entries;
  }

  static DateTime _slotDateTime(DateTime day, String slot) {
    final parts = slot.split(':');
    return DateTime(
      day.year,
      day.month,
      day.day,
      int.parse(parts[0]),
      int.parse(parts[1]),
    );
  }

  static int _parseTime(String time) {
    final parts = time.split(':');
    return int.parse(parts[0]) * 60 + int.parse(parts[1]);
  }

  static String _formatTime(int totalMinutes) {
    final hours = totalMinutes ~/ 60;
    final minutes = totalMinutes % 60;
    return '${hours.toString().padLeft(2, '0')}:${minutes.toString().padLeft(2, '0')}';
  }
}

File: barberia-web/frontend/src/lib/server/utils/timeSlotGenerator.ts
import { SERVICE_DURATION_BLOCK } from '@/lib/server/utils/serviceDuration';
import { expandOccupiedSlots, rangesOverlap, timeToMinutes } from '@/lib/server/utils/appointmentSlots';

export type ScheduleConfig = {
  scheduleStart: string;
  scheduleEnd: string;
  scheduleInterval: number;
};

export type SlotStatus = 'available' | 'pastUnavailable' | 'blocked' | 'booked';

export type SlotEntry = {
  time: string;
  status: SlotStatus;
};

function parseTime(time: string): number {
  const [hours, minutes] = time.split(':').map(Number);
  return hours * 60 + minutes;
}

function formatTime(totalMinutes: number): string {
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`;
}

export function generateAllSlots(config: ScheduleConfig): string[] {
  const slots: string[] = [];
  let current = parseTime(config.scheduleStart);
  const end = parseTime(config.scheduleEnd);
  while (current < end) {
    slots.push(formatTime(current));
    current += config.scheduleInterval;
  }
  return slots;
}

function slotDateTime(day: Date, slot: string): Date {
  const [hours, minutes] = slot.split(':').map(Number);
  return new Date(day.getFullYear(), day.getMonth(), day.getDate(), hours, minutes);
}

function canFitAtStart(
  startTime: string,
  durationMinutes: number,
  scheduleEnd: string,
  occupiedSlots: Set<string>,
  blockedSlots: Set<string>,
): boolean {
  const slots = expandOccupiedSlots(startTime, durationMinutes);
  const endMinutes = timeToMinutes(scheduleEnd);
  for (const slot of slots) {
    if (timeToMinutes(slot) >= endMinutes) return false;
    if (occupiedSlots.has(slot)) return false;
    if (blockedSlots.has(slot)) return false;
  }
  return true;
}

export function buildBookingGrid(params: {
  config: ScheduleConfig;
  date: string;
  occupiedSlots: string[];
  durationMinutes: number;
  blockedTimes?: string[];
  reference?: Date;
  /** Hoy y hora actual en el navegador del cliente (evita UTC en Vercel). */
  localReference?: { date: string; minutesSinceMidnight: number };
}): SlotEntry[] {
  const {
    config,
    date,
    occupiedSlots,
    durationMinutes,
    blockedTimes = [],
    reference = new Date(),
    localReference,
  } = params;

  const day = new Date(`${date}T12:00:00`);
  const referenceDate = localReference
    ? localReference.date
    : (() => {
        const today = new Date(reference.getFullYear(), reference.getMonth(), reference.getDate());
        const y = today.getFullYear();
        const m = String(today.getMonth() + 1).padStart(2, '0');
        const d = String(today.getDate()).padStart(2, '0');
        return `${y}-${m}-${d}`;
      })();
  const isPastDate = date < referenceDate;
  const isToday = date === referenceDate;

  const occupiedSet = new Set(occupiedSlots);
  const blockedSet = new Set(blockedTimes);
  const entries: SlotEntry[] = [];

  let current = parseTime(config.scheduleStart);
  const end = parseTime(config.scheduleEnd);
  const step = Math.min(config.scheduleInterval, SERVICE_DURATION_BLOCK);

  while (current < end) {
    const slot = formatTime(current);
    const isPast = isPastDate
      || (localReference
        ? isToday && current < localReference.minutesSinceMidnight
        : isToday && slotDateTime(day, slot) < reference);
    const slotOccupied = occupiedSet.has(slot);
    const slotBlocked = blockedSet.has(slot);
    const fits = canFitAtStart(
      slot,
      durationMinutes,
      config.scheduleEnd,
      occupiedSet,
      blockedSet,
    );

    if (!fits) {
      if (slotOccupied) {
        entries.push({ time: slot, status: 'booked' });
      } else if (slotBlocked) {
        entries.push({ time: slot, status: 'blocked' });
      } else {
        entries.push({ time: slot, status: 'pastUnavailable' });
      }
    } else if (isPast) {
      entries.push({ time: slot, status: 'pastUnavailable' });
    } else {
      entries.push({ time: slot, status: 'available' });
    }

    current += step;
  }

  return entries;
}

export function occupiedFromAppointments(
  appointments: { time: string; durationMinutes: number }[],
): string[] {
  const occupied = new Set<string>();
  for (const appt of appointments) {
    for (const slot of expandOccupiedSlots(appt.time, appt.durationMinutes)) {
      occupied.add(slot);
    }
  }
  return [...occupied].sort();
}

export { rangesOverlap };
